staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -932,9 +937,16 @@ var (
 		nodeIncludedPredicate,
 		nodeUnTaintedPredicate,
 	}
+	stableNodeSetPredicates []NodeConditionPredicate = []NodeConditionPredicate{
+		nodeIncludedPredicate,
+		nodeUnTaintedPredicate,
```

Comment:
I thought we landed on not syncing load balancers as a consequence of the addition/deletion of this taint, but making sure we never add nodes to the LB set which might have this taint, see: https://github.com/kubernetes/enhancements/pull/3460#discussion_r1095135706 

Logically, what you mentioned is true and possible, though maybe not likely. 

> If CA taints a Node, then deletes the VM, and NodeController is slow (say it takes 60 seconds to wake up), any subsequent LB set changes will include this Node, but the VM for it is already gone, so will fail.

Reply:
> I thought we landed on not syncing load balancers as a consequence of the addition/deletion of this taint, but making sure we never add nodes to the LB set which might have this taint

Agree, but is this what is happening?  The informer setup code filters node updates, but if a node is added or removed, we sync all LBs, and in doing so we get all nodes and then filter by stableNodeSetPredicates, which includes the taint.

So chanign the taint won't trigger a sync, but any node being added or removed would cause all taineted nodes to be removed?  I don't think that's what we want, is it?

Reply:
>  I don't think that's what we want, is it?

Are you saying this because we wouldn't connection drain the tainted nodes? I think we can live with that, simply because I don't think this corner case has a very large window of opportunity to occur....and if it does occur: the effect is that we don't connection drain, which we can live with.   

Beyond what you mentioned: 

> If CA taints a Node, then deletes the VM, and NodeController is slow (say it takes 60 seconds to wake up), any subsequent LB set changes will include this Node, but the VM for it is already gone, so will fail.

the events triggering "subsequent LB set changes", which would be a) the addition of a new Node b) change to `"node.kubernetes.io/exclude-from-external-load-balancers"` c) the deletion of a Node but not by the NodeController (so, only by a user because the node controller is the only thing in Kube deleting nodes, I believe). 
In "KEP-2" kube-proxy will start connection draining on the `unschedulable` condition: reducing the window of opportunity to not connection drain even further. 

Also

>  I mentioned excluding a tainted node if it was not already in the LB set,

Any tainted node will always be in the LB set the moment it is tainted. So if we're only excluding the tainted node if it's not in the LB set, then we're probably never excluding it, right? 

What I am trying to say is: I think not syncing on changes to the taint but excluding all nodes with it, is fine. I am also fine with removing the taint altogether, because I think it's acceptable to have the service controller experience transient errors until this condition resolves. But I can also see the argument of people operating the control plane components and wanting to avoid errors

Reply:
>> I don't think that's what we want, is it?
>
> Are you saying this because we wouldn't connection drain the tainted nodes?

Exactly.  In a cluster that is scaling down via CA, will we only see one of these at a time?  If so, it's probably OK.

My fear is that something like:

* LB set is nodes A, B, C
* CA decides to "rebalance" or make 2 independent decisions.  E.g. "node A is almost empty, scale down" and "there's a pod that needs some special resource which needs a new node"
* taint Node A (but do not remove it) & set unschedulable
* draining period starts
* new node comes up
* node A is removed from LB set

How about this:  Add a big fat comment here, along the lines of:

```
// This is not perfect, but probably good enough.  We won't update the LBs
// just because the taint was added (see shouldSyncUpdatedNode) but if any
// other situation causes an LB sync, tainted nodes will be excluded at that
// time.
```

To do better than that is actually more complicated - just leaving tainted nodes that are already in a set can still fail (as you cited me saying :), and the corner-case is small.  Just leave us a clue so when customers call about occasional errors from LBs, we can hate past-us a little less. :)

Reply:
OK, I'll update with that comment.